Document installation of Microsoft build of Go for Ubuntu (#1804)
microsoft/go: README.md

* In Azure Linux 2.0, the package `msft-golang` installs this fork.
* In Azure Linux 3.0, the `golang` package installs this fork.

### Ubuntu

To install the Microsoft build of Go using an Ubuntu package, first set up the [Linux package repository for Microsoft Products](https://learn.microsoft.com/en-us/linux/packages).
Packages are available in the Ubuntu 22.04 and 24.04 repositories.

> After the repository is added, install the Microsoft build of Go by running the following commands:

Reviewer: ```suggestion

After the repository is added, install the Microsoft build of Go by running the following commands:
```

Missed newline. (So that those who have the repo already set up can easily visually scan to the next step.)

Author: Fixed

* In Azure Linux 2.0, the package `msft-golang` installs this fork.
* In Azure Linux 3.0, the `golang` package installs this fork.

### Ubuntu

> To install the Microsoft Build of Go using an Ubuntu package, open a terminal and run the following commands:

Reviewer: ```suggestion
To install the Microsoft build of Go using an Ubuntu package, open a terminal and run the following commands:
```

(For some reason we have the "Microsoft Build of OpenJDK" and "Microsoft build of Go". 🤷)

```bash
# Supported Ubuntu versions: 22.04 and 24.04.
# For other versions, please use the tar.gz package.

> # See Microsoft's official instructions for up-to-date steps:

Reviewer: IMO we shouldn't maintain our own copy of the repo setup steps at all, only a pointer to the docs, a list of which Ubuntu versions have packages available, and `sudo apt-get update && sudo apt-get install msft-golang`.

# https://learn.microsoft.com/linux/packages
ubuntu_release=`lsb_release -rs`
wget https://packages.microsoft.com/config/ubuntu/${ubuntu_release}/packages-microsoft-prod.deb -O packages-microsoft-prod.deb
sudo dpkg -i packages-microsoft-prod.deb
```

> After the repository is added, install the Microsoft Build of Go by running the following commands:

Reviewer: ```suggestion
To install the Microsoft build of Go using an Ubuntu package, first set up the [Linux package repository for Microsoft Products](https://learn.microsoft.com/en-us/linux/packages).
Packages are available in the Ubuntu 22.04 and 24.04 repositories.

After the repository is added, install the Microsoft build of Go by running the following commands:
```

* In Azure Linux 2.0, the package `msft-golang` installs this fork.
* In Azure Linux 3.0, the `golang` package installs this fork.

### Ubuntu

> To install the Microsoft Build of Go on Ubuntu 22.04 or 24.04 LTS, open a terminal and run the following commands:

Reviewer 2: ```suggestion
To install the Microsoft Build of Go on Ubuntu, open a terminal and run the following commands:
```

Author: Thanks

```bash
# Valid values are only '22.04' and '24.04'
# For other versions of Ubuntu, please use the tar.gz package
ubuntu_release=`lsb_release -rs`
wget https://packages.microsoft.com/config/ubuntu/${ubuntu_release}/packages-microsoft-prod.deb -O packages-microsoft-prod.deb
sudo dpkg -i packages-microsoft-prod.deb
```
After the repository is added, install the Microsoft Build of Go by running the following commands:
```bash
sudo apt-get update

> sudo apt-get install msft-golang

Reviewer 2: perhaps add a note on how to install a specific version?

Author: Sure,added

```

> To verify the installation:

Reviewer 2: This is probably overkill but I'll leave it to @dagood to decide if he thinks it's useful

Author: Okey, keeping this for now.

* In Azure Linux 2.0, the package `msft-golang` installs this fork.
* In Azure Linux 3.0, the `golang` package installs this fork.

### Ubuntu

> To install the Microsoft Build of Go on Ubuntu, open a terminal and run the following commands:

Reviewer: ```suggestion
To install the Microsoft Build of Go using an Ubuntu package, open a terminal and run the following commands:
```

Nitpick, but other methods are also valid on Ubuntu, and this is only useful to those who prefer the central package feed. (I don't tend to prefer Microsoft packages personally because of the setup complication and the risk of adding the Microsoft repository.)

> ```bash

Reviewer 2: I always prefer codeblocks to be wrapped with a linebreak

```suggestion

```bash
```

# Valid values are only '22.04' and '24.04'
# For other versions of Ubuntu, please use the tar.gz package

> ubuntu_release=`lsb_release -rs`

Reviewer: We can improve this in a few ways:

* `set -eu` to exit upon failure and detect unbound vars.
* `lsb_release` doesn't exist by default in some environments, don't use it.
* Check the requirements and tell the user about it. Don't ask the user to do it.

```sh
(
  set -eu
  ubuntu_release=$(. /etc/os-release; echo "$VERSION_ID")
  supported_versions='22.04 24.04'
  for v in $supported_versions; do
    if [ "$v" = "$ubuntu_release" ]; then
      wget "https://packages.microsoft.com/config/ubuntu/${ubuntu_release}/packages-microsoft-prod.deb" -O packages-microsoft-prod.deb
      sudo dpkg -i packages-microsoft-prod.deb
      exit 0
    fi
  done
  echo "Unsupported Ubuntu release: $ubuntu_release. Supported: $supported_versions. Consider using a different install method." >&2
  exit 1
)
```

Every Microsoft team having their own different, loose, instructions really bugs me. 😅

I know it's not typical to run in a subshell and use `set -eu` with a doc snippet, but it makes it a lot more like running a proper script, lets you use `exit`, allows normal safeguards, and avoids leaking env variables to the user's session.

We could consider including a `curl` fallback too, and handling already running as the root user (no `sudo` dependency in that case).

Reviewer 2: I wonder if we should provide a shell script to simplify this process?

e.g

```bash
/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/microsoft/go/HEAD/setup-pmc.sh)"
```

Which would check if `packages.microsoft.com` has been added or not, setup the repo if it's not and provide a few "getting started" hints so people can then just `apt-get install msft-golang`. WFYT?

Reviewer 2: I played around with this idea here: https://github.com/microsoft/go-infra/pull/304

Reviewer: Piping a shell script is one of those "considered dangerous" topics, I don't think we should encourage it. (Especially from `raw.githubusercontent.com`.)

It looks like the packages.microsoft.com team has actually expanded their online presence! We should just link to https://learn.microsoft.com/en-us/linux/packages.

Reviewer: (And if that site doesn't look good enough, we should get them to make it good, IMO.)

Reviewer: > It looks like the packages.microsoft.com team has actually expanded their online presence! We should just link to https://learn.microsoft.com/en-us/linux/packages.

(Highlighting the specific action item here to make it clear because my review was re-requested. 😄)

Pointing out here which distro-versions actually have packages available is still useful, but the instructions PMC maintains should be what people use.

wget https://packages.microsoft.com/config/ubuntu/${ubuntu_release}/packages-microsoft-prod.deb -O packages-microsoft-prod.deb
sudo dpkg -i packages-microsoft-prod.deb

> ```

Reviewer 2: Same as above

```suggestion

```
```

After the repository is added, install the Microsoft Build of Go by running the following commands:
```bash

> sudo apt-get update
> sudo apt-get install msft-golang

Reviewer: Use `&&` to avoid easily missing errors that happen during the update. (Or use a subshell and `set -e` to exit early, but probably overkill for this line.)

```suggestion
sudo apt-get update && sudo apt-get install msft-golang
```

```
Installing a Specific Version

> To install a specific version of the Microsoft Build of Go:

Reviewer: Too much for the readme IMO. None of the other instructions in the readme file include how to get other versions.

(We could have a new file in `eng/docs` that we link to that goes into detail about installing specific Ubuntu packages if we really want to maintain this info.)

List available versions:

```bash
apt-cache madison msft-golang
```
Install a specific version by specifying it explicitly:
```bash
sudo apt-get install msft-golang=<version>
```
For example:
```bash
sudo apt-get install msft-golang=1.22.3-1
```
Replace 1.22.3-1 with the version string listed in the apt-cache madison output.

To verify the installation:
```bash

> go version

Reviewer: IMO this is too much info for the readme.

Also: this `go` could also be unrelated to the package. 😄 If you don't confirm that `go version` fails to find `go` before running the steps, you can't really know where this `go` came from.
